File: android/app/src/main/java/com/example/app/NativeUploaderPlugin.java
package com.example.app;

import android.app.Activity;
import android.content.ClipData;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.Environment;
import android.os.Handler;
import android.os.Looper;
import android.os.SystemClock;
import android.provider.OpenableColumns;

import androidx.activity.result.ActivityResult; // ✅ this is the one you need

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.ActivityCallback;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@CapacitorPlugin(name = "NativeUploader")
public class NativeUploaderPlugin extends Plugin {

    private final ExecutorService ioExecutor = Executors.newSingleThreadExecutor();
    private final Handler main = new Handler(Looper.getMainLooper());

    @PluginMethod
    public void pickAndStageMany(PluginCall call) {
        call.setKeepAlive(true);

        Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        intent.setType("*/*");
        intent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true);

        startActivityForResult(call, intent, "onPickedFiles");
    }

    @PluginMethod
    public void deleteFile(PluginCall call) {
        String absolutePath = call.getString("absolutePath");
        if (absolutePath == null || absolutePath.isEmpty()) {
            call.reject("absolutePath is required");
            return;
        }

        ioExecutor.execute(() -> {
            try {
                File file = new File(absolutePath);
                if (!file.exists()) {
                    rejectOnMain(call, "File does not exist: " + absolutePath);
                    return;
                }

                boolean deleted = file.delete();
                if (deleted) {
                    resolveOnMain(call, new JSObject());
                } else {
                    rejectOnMain(call, "Failed to delete file: " + absolutePath);
                }
            } catch (Exception e) {
                rejectOnMain(call, "Error deleting file: " + e.getMessage());
            }
        });
    }

    // ✅ correct signature for Capacitor v3+
    @ActivityCallback
    private void onPickedFiles(PluginCall call, ActivityResult result) {
        int resultCode = result.getResultCode();
        Intent data = result.getData();

        if (resultCode == Activity.RESULT_CANCELED) {
            rejectOnMain(call, "User cancelled");
            return;
        }

        if (data == null) {
            rejectOnMain(call, "No intent data returned");
            return;
        }

        int maxFiles = 2;
        try {
            Integer mf = call.getInt("maxFiles");
            if (mf != null) maxFiles = mf;
        } catch (Exception ignored) {}
        if (maxFiles < 1) maxFiles = 1;
        if (maxFiles > 2) maxFiles = 2;

        List<Uri> uris = new ArrayList<>();
        ClipData clipData = data.getClipData();

        if (clipData != null) {
            for (int i = 0; i < clipData.getItemCount() && uris.size() < maxFiles; i++) {
                Uri u = clipData.getItemAt(i).getUri();
                if (u != null) uris.add(u);
            }
        } else {
            Uri u = data.getData();
            if (u != null) uris.add(u);
        }

        if (uris.isEmpty()) {
            rejectOnMain(call, "No file(s) selected");
            return;
        }

        final List<Uri> finalUris = uris;

        ioExecutor.execute(() -> {
            try {
                // Use external files directory with documents subdirectory (accessible via file manager)
                File docsRoot = getContext().getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS);
                if (docsRoot == null) docsRoot = getContext().getFilesDir();

                File destDir = new File(docsRoot, "HSC-SESSIONS/FILES");
                //noinspection ResultOfMethodCallIgnored
                destDir.mkdirs();

                JSArray results = new JSArray();
                long stamp = System.currentTimeMillis();

                for (int idx = 0; idx < finalUris.size(); idx++) {
                    Uri uri = finalUris.get(idx);

                    String originalName = getDisplayName(uri);
                    if (originalName == null || originalName.trim().isEmpty()) {
                        originalName = "upload_" + stamp + "_" + idx;
                    }

                    String safeName = originalName.replaceAll("[^a-zA-Z0-9._-]", "_");
                    String stampedName = stamp + "_" + idx + "_" + safeName;

                    File partial = new File(destDir, stampedName + ".partial");
                    File finalFile = new File(destDir, stampedName);

                    long expectedSize = getSize(uri); // -1 if unknown
                    long written = 0L;
                    long lastEmitMs = 0L;

                    try (InputStream in = getContext().getContentResolver().openInputStream(uri);
                         FileOutputStream out = new FileOutputStream(partial)) {

                        if (in == null) throw new IllegalStateException("Unable to open input stream");

                        byte[] buf = new byte[1024 * 1024]; // 1MB chunks
                        int read;
                        while ((read = in.read(buf)) != -1) {
                            out.write(buf, 0, read);
                            written += read;

                            long nowMs = SystemClock.uptimeMillis();
                            if (nowMs - lastEmitMs >= 250) {
                                lastEmitMs = nowMs;
                                JSObject ev = new JSObject();
                                ev.put("fileIndex", idx);
                                ev.put("bytesWritten", written);
                                ev.put("totalBytes", expectedSize);
                                ev.put("originalName", originalName);
                                notifyOnMain("uploadProgress", ev);
                            }
                        }
                        out.flush();
                    }

                    if (finalFile.exists()) {
                        //noinspection ResultOfMethodCallIgnored
                        finalFile.delete();
                    }
                    if (!partial.renameTo(finalFile)) {
                        throw new IllegalStateException("Rename failed");
                    }

                    String mimeType = getContext().getContentResolver().getType(uri);
                    if (mimeType == null) mimeType = "application/octet-stream";

                    JSObject one = new JSObject();
                    one.put("absolutePath", finalFile.getAbsolutePath());
                    one.put("logicalPath", "DOCUMENTS/HSC-SESSIONS/FILES/" + finalFile.getName());
                    one.put("size", finalFile.length());
                    one.put("mimeType", mimeType);
                    one.put("status", "staged");
                    one.put("originalName", originalName);

                    results.put(one);
                }

                JSObject ret = new JSObject();
                ret.put("files", results);
                resolveOnMain(call, ret);

            } catch (Exception e) {
                rejectOnMain(call, "Stage failed: " + e.getMessage());
            }
        });
    }

    private void resolveOnMain(PluginCall call, JSObject ret) {
        main.post(() -> {
            try { call.resolve(ret); }
            finally { call.setKeepAlive(false); }
        });
    }

    private void rejectOnMain(PluginCall call, String msg) {
        main.post(() -> {
            try { call.reject(msg); }
            finally { call.setKeepAlive(false); }
        });
    }

    private void notifyOnMain(String eventName, JSObject data) {
        main.post(() -> notifyListeners(eventName, data));
    }

    private String getDisplayName(Uri uri) {
        Cursor cursor = null;
        try {
            cursor = getContext().getContentResolver().query(uri, null, null, null, null);
            if (cursor != null && cursor.moveToFirst()) {
                int idx = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME);
                if (idx >= 0) return cursor.getString(idx);
            }
        } catch (Exception ignored) {
        } finally {
            if (cursor != null) cursor.close();
        }
        return null;
    }

    private long getSize(Uri uri) {
        Cursor cursor = null;
        try {
            cursor = getContext().getContentResolver().query(uri, null, null, null, null);
            if (cursor != null && cursor.moveToFirst()) {
                int idx = cursor.getColumnIndex(OpenableColumns.SIZE);
                if (idx >= 0) return cursor.getLong(idx);
            }
        } catch (Exception ignored) {
        } finally {
            if (cursor != null) cursor.close();
        }
        return -1L;
    }
}

File: kt-msca-plugins/NativeUploaderPlugin.kt
package org.deal.mcsa.plugins

import android.app.Activity
import android.content.ClipData
import android.content.Intent
import android.net.Uri
import android.os.Environment
import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.provider.OpenableColumns
import android.util.Base64
import androidx.activity.result.ActivityResult
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.ActivityCallback
import com.getcapacitor.annotation.CapacitorPlugin
import java.io.File
import java.io.FileOutputStream
import java.util.concurrent.Executors

@CapacitorPlugin(name = "NativeUploader")
class NativeUploaderPlugin : Plugin() {

    private val ioExecutor = Executors.newSingleThreadExecutor()
    private val main = Handler(Looper.getMainLooper())

    @PluginMethod
    fun pickAndStageMany(call: PluginCall) {
        call.setKeepAlive(true)

        val intent = Intent(Intent.ACTION_GET_CONTENT).apply {
            addCategory(Intent.CATEGORY_OPENABLE)
            type = "*/*"
            putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true)
        }

        startActivityForResult(call, intent, "onPickedFiles")
    }

    @PluginMethod
    fun deleteFile(call: PluginCall) {
        val absolutePath = call.getString("absolutePath")
        if (absolutePath.isNullOrEmpty()) {
            call.reject("absolutePath is required")
            return
        }

        ioExecutor.execute {
            try {
                val file = File(absolutePath)
                if (!file.exists()) {
                    rejectOnMain(call, "File does not exist: $absolutePath")
                    return@execute
                }

                val deleted = file.delete()
                if (deleted) {
                    resolveOnMain(call, JSObject())
                } else {
                    rejectOnMain(call, "Failed to delete file: $absolutePath")
                }
            } catch (e: Exception) {
                rejectOnMain(call, "Error deleting file: ${e.message}")
            }
        }
    }

    @PluginMethod
    fun saveExtractedFile(call: PluginCall) {
        val base64Data = call.getString("base64Data")
        val fileName = call.getString("fileName")
        
        if (base64Data.isNullOrEmpty()) {
            call.reject("base64Data is required")
            return
        }
        if (fileName.isNullOrEmpty()) {
            call.reject("fileName is required")
            return
        }

        ioExecutor.execute {
            try {
                // Use the same directory as pickAndStageMany
                var docsRoot = context.getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS)
                if (docsRoot == null) docsRoot = context.filesDir

                val destDir = File(docsRoot, "HSC-SESSIONS/FILES")
                destDir.mkdirs()

                val finalFile = File(destDir, fileName)

                // Decode base64 and write to file
                val fileData = Base64.decode(base64Data, Base64.DEFAULT)
                FileOutputStream(finalFile).use { out ->
                    out.write(fileData)
                    out.flush()
                }

                var mimeType = call.getString("mimeType")
                if (mimeType == null) mimeType = "application/octet-stream"

                val result = JSObject()
                result.put("absolutePath", finalFile.absolutePath)
                result.put("logicalPath", "DOCUMENTS/HSC-SESSIONS/FILES/${finalFile.name}")
                result.put("size", finalFile.length())
                result.put("mimeType", mimeType)

                resolveOnMain(call, result)
            } catch (e: Exception) {
                rejectOnMain(call, "Save failed: ${e.message}")
            }
        }
    }

    @ActivityCallback
    private fun onPickedFiles(call: PluginCall, result: ActivityResult) {
        val resultCode = result.resultCode
        val data = result.data

        if (resultCode == Activity.RESULT_CANCELED) {
            rejectOnMain(call, "User cancelled")
            return
        }

        if (data == null) {
            rejectOnMain(call, "No intent data returned")
            return
        }

        var maxFiles = call.getInt("maxFiles") ?: 2
        if (maxFiles < 1) maxFiles = 1
        if (maxFiles > 2) maxFiles = 2

        val uris = mutableListOf<Uri>()
        val clipData = data.clipData

        if (clipData != null) {
            for (i in 0 until clipData.itemCount) {
                if (uris.size >= maxFiles) break
                clipData.getItemAt(i).uri?.let { uris.add(it) }
            }
        } else {
            data.data?.let { uris.add(it) }
        }

        if (uris.isEmpty()) {
            rejectOnMain(call, "No file(s) selected")
            return
        }

        val finalUris = uris.toList()

        ioExecutor.execute {
            try {
                var docsRoot = context.getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS)
                if (docsRoot == null) docsRoot = context.filesDir

                val destDir = File(docsRoot, "HSC-SESSIONS/FILES")
                destDir.mkdirs()

                val results = JSArray()
                val stamp = System.currentTimeMillis()

                for ((idx, uri) in finalUris.withIndex()) {
                    var originalName = getDisplayName(uri)
                    if (originalName.isNullOrBlank()) {
                        originalName = "upload_${stamp}_$idx"
                    }

                    // Note: File extension validation is done in JavaScript to show proper error messages
                    // All files are staged here, validation happens later in handleUpload

                    val safeName = originalName.replace(Regex("[^a-zA-Z0-9._-]"), "_")
                    val stampedName = "${stamp}_${idx}_$safeName"

                    val partial = File(destDir, "$stampedName.partial")
                    val finalFile = File(destDir, stampedName)

                    val expectedSize = getSize(uri)
                    var written = 0L
                    var lastEmitMs = 0L

                    context.contentResolver.openInputStream(uri)?.use { input ->
                        FileOutputStream(partial).use { out ->
                            val buf = ByteArray(1024 * 1024) // 1MB chunks
                            var read: Int
                            while (input.read(buf).also { read = it } != -1) {
                                out.write(buf, 0, read)
                                written += read

                                val nowMs = SystemClock.uptimeMillis()
                                if (nowMs - lastEmitMs >= 250) {
                                    lastEmitMs = nowMs
                                    val ev = JSObject()
                                    ev.put("fileIndex", idx)
                                    ev.put("bytesWritten", written)
                                    ev.put("totalBytes", expectedSize)
                                    ev.put("originalName", originalName)
                                    notifyOnMain("uploadProgress", ev)
                                }
                            }
                            out.flush()
                        }
                    } ?: throw IllegalStateException("Unable to open input stream")

                    if (finalFile.exists()) {
                        finalFile.delete()
                    }
                    if (!partial.renameTo(finalFile)) {
                        throw IllegalStateException("Rename failed")
                    }

                    var mimeType = context.contentResolver.getType(uri)
                    if (mimeType == null) mimeType = "application/octet-stream"

                    val one = JSObject()
                    one.put("absolutePath", finalFile.absolutePath)
                    one.put("logicalPath", "DOCUMENTS/HSC-SESSIONS/FILES/${finalFile.name}")
                    one.put("size", finalFile.length())
                    one.put("mimeType", mimeType)
                    one.put("status", "staged")
                    one.put("originalName", originalName)

                    results.put(one)
                }

                val ret = JSObject()
                ret.put("files", results)
                resolveOnMain(call, ret)

            } catch (e: Exception) {
                rejectOnMain(call, "Stage failed: ${e.message}")
            }
        }
    }

    private fun resolveOnMain(call: PluginCall, ret: JSObject) {
        main.post {
            try {
                call.resolve(ret)
            } finally {
                call.setKeepAlive(false)
            }
        }
    }

    private fun rejectOnMain(call: PluginCall, msg: String) {
        main.post {
            try {
                call.reject(msg)
            } finally {
                call.setKeepAlive(false)
            }
        }
    }

    private fun notifyOnMain(eventName: String, data: JSObject) {
        main.post { notifyListeners(eventName, data) }
    }

    private fun getDisplayName(uri: Uri): String? {
        return try {
            context.contentResolver.query(uri, null, null, null, null)?.use { cursor ->
                if (cursor.moveToFirst()) {
                    val idx = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                    if (idx >= 0) cursor.getString(idx) else null
                } else null
            }
        } catch (e: Exception) {
            null
        }
    }

    private fun getSize(uri: Uri): Long {
        return try {
            context.contentResolver.query(uri, null, null, null, null)?.use { cursor ->
                if (cursor.moveToFirst()) {
                    val idx = cursor.getColumnIndex(OpenableColumns.SIZE)
                    if (idx >= 0) cursor.getLong(idx) else -1L
                } else -1L
            } ?: -1L
        } catch (e: Exception) {
            -1L
        }
    }
}

